### PR TITLE
Make hot restart tolerate premature death of the parent instance

### PR DESCRIPTION
**Pull request by ravenblackx**
Commit Message: Make hot restart tolerate premature death of the parent instance
Additional Description: The current behavior of hot restart makes it difficult to manage crashes, especially when envoy instances are run in containers - if epoch 1 crashed before epoch 2 tries to start up, there's no simple mechanism to have epoch 2 detect the crash and run as epoch 0 instead. If epoch 2 tries to start up while epoch 1 is not running, epoch 2 also crashes during startup, potentially causing an endless crashloop if there isn't some sort of "break out" detection.
This change makes it so instead of crashing during startup if the parent instance is not responsive, the child instance does one check at the beginning of its initialization of whether the parent is responsive, and if it is not, the child starts up without the hot restart socket/stats handoff, as if it was epoch 0 (but keeps the epoch, so any manager doesn't need to accommodate this).
Risk Level: Very small - new behavior is command-line-flag guarded and default is to persist the old behavior.
Testing: Unit tests and also added to the hot restart end-to-end test. Also patched in and deployed in production, which resolved the fact that we couldn't get hot restart to play nicely with our production tooling any other way!
Docs Changes: Documented the new flag.
Release Notes: Linked to the new flag documentation.
Platform Specific Features: n/a

### CODE REVIEW

**Comment by soulxu** [comment]
> Commit Message: Make hot restart tolerate premature death of the parent instance Additional Description: The current behavior of hot restart makes it difficult to manage crashes, especially when envoy instances are run in containers

I found the `hot-restarter.py` handling the similar thing (if I understand correctly)

https://github.com/envoyproxy/envoy/blob/d4697a86717d8f2adc8bc825e20644c66437a7be/docs/root/operations/hot_restarter.rst?plain=1#L45-L47

So a question is whether the people will use hot restart feature without a script like `hot-restarter.py` in container, if yes, then this option is useful.

**Comment by ravenblackx** [comment]
> So a question is whether the people will use hot restart feature without a script like `hot-restarter.py` in container, if yes, then this option is useful.

Yes, hot_restarter.py doesn't work for us because we have a whole deployment management system that can't be adjusted to work with its requirements (the expectation of the script being that it *is not* shut down for a new version but *is* shut down if it crashed, would mean the deployment management would have to be able to treat the two situations differently, which would be a quite unique requirement). It's *possible* the script could be modified for our use-case to treat the signals that come from the deployment management system differently in a way that would make it work, but after having tried to get any sort of wrapper to work correctly for quite a while, this change is *so simple* by comparison (and perhaps it could also be used to simplify hot_restarter.py)

File: docs/root/operations/cli.rst
Comment (soulxu) [inline docs/root/operations/cli.rst:70]: is it worth a release note?

File: docs/root/operations/cli.rst
Comment (soulxu) [inline docs/root/operations/cli.rst:74]: maybe clarify that this is only available when the hot restart initialize stage. If the parent was terminated after hot restart, this won't help.

**Review by soulxu — APPROVED** [review]
general LGTM, few doc comments, thanks!

File: docs/root/operations/cli.rst
Comment (ravenblackx) [reply docs/root/operations/cli.rst:74]: Done.

File: docs/root/operations/cli.rst
Comment (ravenblackx) [reply docs/root/operations/cli.rst:70]: Added a release note. I don't know if it's worth it, I think using hot restart is extremely niche, but no harm!

**Review by soulxu — APPROVED** [review]
LGTM, thanks!

**Comment by soulxu** [comment]
cc @markdroth for senior maintainer review

**Comment by RyanTheOptimist** [comment]
@markdroth: friendly maintainer-on-call ping

**Comment by markdroth** [comment]
I think this particular PR shouldn't actually require API approval, since it's affecting only Envoy's admin interface.  I'll go ahead and mark it as API spprtoved to unblock it, but one of the Envoy maintainers will need to make the final decision here.

/lgtm api

**Comment by ravenblackx** [comment]
+yanavlasov for final senior maintainer pass.

**Comment by lambdai** [comment]
> This change makes it so instead of crashing during startup if the parent instance is not responsive, the child instance does one check at the beginning of its initialization of whether the parent is responsive, and if it is not, the child starts up without the hot restart socket/stats handoff, as if it was epoch 0 (but keeps the epoch, so any manager doesn't need to accommodate this).

Aha, at our side, the previous process death is detected by the bootstrap manager. It's nice to have it as envoy build-in functionality

**Comment by ravenblackx** [comment]
> Aha, at our side, the previous process death is detected by the bootstrap manager. It's nice to have it as envoy build-in functionality

I tried to do that but with containerization it was somewhere between impossible and very difficult. (Detecting whether the PID is running would theoretically have been a pretty reliable method, but with running in separate containers the PIDs overlap and also can't see each other. Doing guards with the shared filesystem is always just a danger of badly timed crashes leaving unusable state, much like #33446 is about.)

*Ideally* this change would also accommodate crashes part way through initialization, but getting at least the common case resolved is a massive improvement already.

**Review by yanavlasov — APPROVED** [review]

**Pull request merged**
